### PR TITLE
SC : Add extended bridge contract / callback contract example

### PR DESCRIPTION
**Pull request by KimKyungup**
## Proposed changes
This PR added extended bridge / callback contract example and unit test.
Service provider can design their own customized bridge contract and callback contract for their own feature.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments
After this PR, extraData field in request/handleValueTransfer method, will be added after all previous  are merged.

### CODE REVIEW

File: contracts/bridge_test.go [hunk 1]
```diff
@@ -318,16 +330,12 @@ func TestBridgePublicVariables(t *testing.T) {
 	chargeAmount := big.NewInt(10000000)
 	bridgeAccount.Value = chargeAmount
 	bridgeAddress, tx, b, err := bridge.DeployBridge(bridgeAccount, backend, false)
-	if err != nil {
-		t.Fatalf("fail to DeployBridge %v", err)
-	}
+	assert.Equal(t, nil, err)
```
Comment (laggu) [inline contracts/bridge_test.go]: what about using assert.NoError(t, err)?

File: contracts/bridge_test.go [hunk 2]
```diff
@@ -318,16 +330,12 @@ func TestBridgePublicVariables(t *testing.T) {
...
 	backend.Commit()
-	WaitMined(tx, backend, t)
+	assert.Equal(t, nil, WaitMined(tx, backend, t))
```
Comment (laggu) [inline contracts/bridge_test.go]: what about using assert.Nil(t, WaitMined(tx, backend, t))

File: contracts/bridge_test.go [hunk 3]
```diff
@@ -371,3 +379,95 @@ func TestBridgePublicVariables(t *testing.T) {
...
+	registerOfferEventCh := make(chan *extbridge.CallbackRegisteredOffer, 10)
+	registerOfferEventSub, err := cb.WatchRegisteredOffer(nil, registerOfferEventCh)
+	assert.Equal(t, nil, err)
+	defer registerOfferEventSub.Unsubscribe()
+
+	// Subscribe bridge contract event
+	b, err := bridge.NewBridge(bridgeAddr, backend) // create base bridge contract object, not extBridge object
+	assert.Equal(t, nil, err)
+	handleValueTransferEventCh := make(chan *bridge.BridgeHandleValueTransfer, 10)
+	handleSub, err := b.WatchHandleValueTransfer(nil, handleValueTransferEventCh)
```
Comment (laggu) [inline contracts/bridge_test.go]: Is it okay to pass checking error?

File: contracts/bridge_test.go (hunk 1 above)
Comment (KimKyungup) [reply contracts/bridge_test.go]: I will make a PR for all code. :) Thanks.

Comment (KimKyungup) [reply contracts/bridge_test.go]: Anyway, I reflect to this code. kaka

File: contracts/bridge_test.go (hunk 3 above)
Comment (KimKyungup) [reply contracts/bridge_test.go]: I added err check. :)

File: contracts/extbridge/ext_bridge.sol [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        callback = _addr;
+    }
+
+    // handleERC20Transfer sends the ERC20 token by the request and processes extended feature.
+    function handleERC20Transfer(
+        uint256 _amount,
+        address _to,
+        address _contractAddress,
+        uint64 _requestNonce,
+        uint64 _requestBlockNumber//,
```
Comment (laggu) [inline contracts/extbridge/ext_bridge.sol]: The comment seems unnecessary.

Comment (KimKyungup) [reply contracts/extbridge/ext_bridge.sol]: I will uncomment in next PR. So I wrote it. :)

File: contracts/generate.go [hunk 5]
```diff
@@ -17,6 +17,9 @@
 package contracts
 
 //go:generate abigen --sol ./bridge/Bridge.sol --pkg bridge --out ./bridge/Bridge.go
+//go:generate abigen --sol ./extbridge/ext_bridge.sol --pkg extbridge --out ./extbridge/ext_bridge.go
```
Comment (jimni1222) [inline contracts/generate.go]: Can i ask what is this for?

Comment (KimKyungup) [reply contracts/generate.go]: This code make us to generate go binding code of the solidity file like below.

```
➜  klaytn git:(190723e_refactor_contract_method_parameter) cd contracts 
➜  contracts git:(190723e_refactor_contract_method_parameter) go generate
```

And then you can find the genearted file like below.
`./extbridge/ext_bridge.go`

**Comment by KimKyungup** [comment]
@jeonghonah PTAL. :)

File: contracts/bridge_test.go [hunk 6]
```diff
@@ -49,6 +52,15 @@ func WaitMined(tx *types.Transaction, backend bind.DeployBackend, t *testing.T)
 		t.Fatal("Failed to WaitMined.", "err", err, "txHash", tx.Hash().String(), "status", receipt.Status)
 		return err
 	}
+
+	if receipt == nil {
+		return errors.New("not found receipt")
```
Comment (jeonghonah) [inline contracts/bridge_test.go]:
```suggestion
		return errors.New("receipt not found")
```

File: contracts/bridge_test.go [hunk 7]
```diff
@@ -371,3 +380,96 @@ func TestBridgePublicVariables(t *testing.T) {
 	rnonce, err := b.RequestNonce(nil)
 	assert.Equal(t, uint64(0), rnonce)
 }
+
+// TestExtendedBridgeAndCallback checks the following:
+// - the extBridge can make contract call to callback contract.
```
Comment (jeonghonah) [inline contracts/bridge_test.go]:
```suggestion
// - the extBridge can call to a callback contract method.
```

File: contracts/bridge_test.go [hunk 8]
```diff
@@ -371,3 +380,96 @@ func TestBridgePublicVariables(t *testing.T) {
...
+func TestExtendedBridgeAndCallback(t *testing.T) {
+	bridgeAccountKey, _ := crypto.GenerateKey()
+	bridgeAccount := bind.NewKeyedTransactor(bridgeAccountKey)
+
+	//aliceKey, _ := crypto.GenerateKey()
```
Comment (jeonghonah) [inline contracts/bridge_test.go]:
```suggestion
```

File: contracts/bridge_test.go [hunk 9]
```diff
@@ -371,3 +380,96 @@ func TestBridgePublicVariables(t *testing.T) {
...
+	//aliceAcc := bind.NewKeyedTransactor(aliceKey)
```
Comment (jeonghonah) [inline contracts/bridge_test.go]:
```suggestion
```

File: contracts/bridge_test.go [hunk 10]
```diff
@@ -371,3 +380,96 @@ func TestBridgePublicVariables(t *testing.T) {
...
+	backend := backends.NewSimulatedBackend(alloc)
+
+	// Deploy extBridge
+	bridgeAddr, tx, eb, err := extbridge.DeployExtBridge(bridgeAccount, backend, true)
+	assert.NoError(t, err)
+	backend.Commit()
+	assert.Nil(t, WaitMined(tx, backend, t))
+
+	// Deploy token
+	erc20Addr, tx, erc20, err := sctoken.DeployServiceChainToken(bridgeAccount, backend, bridgeAddr)
```
Comment (jeonghonah) [inline contracts/bridge_test.go]: In the next PR, how about changing the `sctoken` to `scerc20`?

File: contracts/bridge_test.go [hunk 11]
```diff
@@ -371,3 +380,96 @@ func TestBridgePublicVariables(t *testing.T) {
...
+	assert.NoError(t, err)
+	backend.Commit()
+	assert.Nil(t, WaitMined(tx, backend, t))
+
+	// Give minter role to bridge contract
+	tx, err = erc20.AddMinter(bridgeAccount, bridgeAddr)
```
Comment (jeonghonah) [inline contracts/bridge_test.go]: Can owner be minter automatically while construction?

File: contracts/bridge_test.go [hunk 12]
```diff
@@ -371,3 +380,96 @@ func TestBridgePublicVariables(t *testing.T) {
...
+	// Check handle request event
+	select {
+	case ev := <-handleValueTransferEventCh:
+		assert.Equal(t, amount.String(), ev.Value.String())
+		assert.Equal(t, rNonce, ev.HandleNonce)
+		assert.Equal(t, erc20Addr, ev.ContractAddress)
+		assert.Equal(t, sc.ERC20, ev.Kind)
+		assert.Equal(t, callbackAddr, ev.Owner)
+
+	case <-time.After(1 * time.Second):
```
Comment (jeonghonah) [inline contracts/bridge_test.go]:
```suggestion
	case <-time.After(time.Second):
```

File: contracts/bridge_test.go [hunk 13]
```diff
@@ -371,3 +380,96 @@ func TestBridgePublicVariables(t *testing.T) {
...
+	}
+
+	// Check Callback event
+	select {
+	case ev := <-registerOfferEventCh:
+		assert.Equal(t, amount.String(), ev.ValueOrID.String())
+		assert.Equal(t, big.NewInt(1), ev.Price) // TODO-Klaytn-Servicechain : it will be tested later.
+		assert.Equal(t, erc20Addr, ev.TokenAddress)
+
+	case <-time.After(1 * time.Second):
```
Comment (jeonghonah) [inline contracts/bridge_test.go]:
```suggestion
	case <-time.After(time.Second):
```

File: contracts/extbridge/callback.sol [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+contract Callback {
+    event RegisteredOffer(
+        address owner,
+        uint256 valueOrID,
+        address tokenAddress,
+        uint256 price
+    );
+
+    function RegisterOffer(
+        address _owner,
```
Comment (jeonghonah) [inline contracts/extbridge/callback.sol]: Just a question:
Why do we need underscore here?

File: contracts/extbridge/ext_bridge.sol [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+import "../bridge/Bridge.sol";
+import "./callback.sol";
+
+contract ExtBridge is Bridge {
+    address public callback = address(0);
+
+    constructor (bool _modeMintBurn) Bridge(_modeMintBurn) public payable {
+    }
+
+    function setCallback(address _addr) public {
```
Comment (jeonghonah) [inline contracts/extbridge/ext_bridge.sol]:
```suggestion
    function setCallback(address _addr) public onlyOnwer {
```

File: contracts/extbridge/ext_bridge.sol [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        callback = _addr;
+    }
+
+    // handleERC20Transfer sends the ERC20 token by the request and processes extended feature.
```
Comment (jeonghonah) [inline contracts/extbridge/ext_bridge.sol]:
```suggestion
    // handleERC20Transfer sends the ERC20 token by the request and processes an extended feature.
```

File: contracts/extbridge/ext_bridge.sol (hunk 4 above)
Comment (jeonghonah) [inline contracts/extbridge/ext_bridge.sol]:
```suggestion
        uint64 _requestBlockNumber,
```

File: contracts/bridge_test.go (hunk 8 above)
Comment (KimKyungup) [reply contracts/bridge_test.go]: This will be used in next PR. :)

File: contracts/bridge_test.go (hunk 9 above)
Comment (KimKyungup) [reply contracts/bridge_test.go]: This will be used in next PR. :)

File: contracts/bridge_test.go (hunk 10 above)
Comment (KimKyungup) [reply contracts/bridge_test.go]: Right. I will do that. :)

File: contracts/bridge_test.go (hunk 11 above)
Comment (KimKyungup) [reply contracts/bridge_test.go]: Yes we can do that if we modifiy sctoken code. I will try it. :)

File: contracts/extbridge/callback.sol (hunk 14 above)
Comment (KimKyungup) [reply contracts/extbridge/callback.sol]: I wrote like this caused by this reason (https://ethereum.stackexchange.com/questions/1351/what-is-the-purpose-of-underscores-in-front-of-parameters).

But It is not formal solidity guide. I will think about it again. And if we need to modify, I will apply the right style for all solidity code once. :)

File: contracts/extbridge/ext_bridge.sol (hunk 4 above)
Comment (KimKyungup) [reply contracts/extbridge/ext_bridge.sol]: I will uncomment it in next PR.

**Comment by KimKyungup** [comment]
@laggu @jimni1222 PTAL again. =) Thanks.

**Pull request merged**